### PR TITLE
Reduce client networking build dependencies.

### PR DESCRIPTION
**Pull request by LGM-Doyle**
This reduces the dependency of compilation units downstream of ClientNetworking.h, improving incremental build times by a factor of 4 when ClientNetworking is changed.  Sadly, it make no difference in the total clean build time.

It moves several dependencies from the header into the cpp file.

It changes function signatures so that forward declarations are as good as full definitions.

It reduces the number of files that include ClientNetworking.h.

I wish I had done this before I started hacking on ClientNetworking.

### CODE REVIEW

**Comment by adrianbroher** [comment]
@LGM-Doyle I'm curious: did you recieve a failing message from AppVeyor via eMail?

File: network/ClientNetworking.cpp [hunk 1]
```diff
@@ -119,10 +134,125 @@ namespace {
...
+    void HandleConnection(boost::asio::ip::tcp::resolver::iterator* it,
+                          const boost::system::error_code& error);
+    void CancelRetries();
+    void NetworkingThread(const std::shared_ptr<const ClientNetworking> self);
+    void HandleMessageBodyRead(const std::shared_ptr<const ClientNetworking>& keep_alive,
+                               boost::system::error_code error, std::size_t bytes_transferred);
+    void HandleMessageHeaderRead(const std::shared_ptr<const ClientNetworking>& keep_alive,
+                                 boost::system::error_code error, std::size_t bytes_transferred);
+    void AsyncReadMessage(const std::shared_ptr<const ClientNetworking>& keep_alive);
+    void HandleMessageWrite(        boost::system::error_code error, std::size_t bytes_transferred);
```
Comment (geoffthemedio) [inline network/ClientNetworking.cpp]: funky indenting

File: network/ClientNetworking.cpp [hunk 2]
```diff
@@ -132,47 +262,53 @@ ClientNetworking::ClientNetworking() :
...
 
-ClientNetworking::ServerList ClientNetworking::DiscoverLANServers() {
+ClientNetworking::ServerNames ClientNetworking::Impl::DiscoverLANServerNames() {
     if (!IsConnected())
-        return ServerList();
+        return ServerNames();
     ServerDiscoverer discoverer(m_io_service);
     discoverer.DiscoverServers();
-    return discoverer.Servers();
+    auto servers = discoverer.Servers();
```
Comment (geoffthemedio) [inline network/ClientNetworking.cpp]: necessity of local copy instead of putting call in subsequent loop directly?

File: network/ClientNetworking.cpp [hunk 3]
```diff
@@ -261,14 +396,15 @@ bool ClientNetworking::ConnectToServer(
...
+bool ClientNetworking::Impl::ConnectToLocalHostServer(
+    const ClientNetworking* const self,
+    const std::chrono::milliseconds& timeout/* = std::chrono::seconds(10)*/)
 {
     bool retval = false;
 #if FREEORION_WIN32
     try {
 #endif
-        retval = ConnectToServer("127.0.0.1", timeout);
+        retval = ConnectToServer(self, "127.0.0.1", timeout);
```
Comment (geoffthemedio) [inline network/ClientNetworking.cpp]: self vs localhost ?

File: network/ClientNetworking.cpp [hunk 4]
```diff
@@ -319,7 +455,7 @@ void ClientNetworking::GetMessage(Message& message) {
                       << message;
 }
 
-void ClientNetworking::SendSynchronousMessage(Message message, Message& response_message) {
+void ClientNetworking::Impl::SendSynchronousMessage(const Message& message, Message& response_message) {
     if (TRACE_EXECUTION)
         DebugLogger() << "ClientNetworking::SendSynchronousMessage : sending message "
                                << message;
```
Comment (geoffthemedio) [inline network/ClientNetworking.cpp]: not this patch, but indenting odd

File: network/ClientNetworking.cpp [hunk 5]
```diff
@@ -331,7 +467,7 @@ void ClientNetworking::SendSynchronousMessage(Message message, Message& response
                                << "response message " << response_message;
 }
 
-void ClientNetworking::HandleConnection(tcp::resolver::iterator* it,
+void ClientNetworking::Impl::HandleConnection(tcp::resolver::iterator* it,
                                         const boost::system::error_code& error)
```
Comment (geoffthemedio) [inline network/ClientNetworking.cpp]: indenting

File: network/ClientNetworking.cpp [hunk 6]
```diff
@@ -387,7 +523,7 @@ void ClientNetworking::NetworkingThread(std::shared_ptr<ClientNetworking>& self)
         DebugLogger() << "ClientNetworking::NetworkingThread() : Networking thread terminated.";
 }
 
-void ClientNetworking::HandleMessageBodyRead(const std::shared_ptr<ClientNetworking>& keep_alive,
+void ClientNetworking::Impl::HandleMessageBodyRead(const std::shared_ptr<const ClientNetworking>& keep_alive,
                                              boost::system::error_code error, std::size_t bytes_transferred)
```
Comment (geoffthemedio) [inline network/ClientNetworking.cpp]: ...

File: network/ClientNetworking.cpp [hunk 7]
```diff
@@ -400,7 +536,7 @@ void ClientNetworking::HandleMessageBodyRead(const std::shared_ptr<ClientNetwork
     }
 }
 
-void ClientNetworking::HandleMessageHeaderRead(const std::shared_ptr<ClientNetworking>& keep_alive,
+void ClientNetworking::Impl::HandleMessageHeaderRead(const std::shared_ptr<const ClientNetworking>& keep_alive,
                                                boost::system::error_code error, std::size_t bytes_transferred)
```
Comment (geoffthemedio) [inline network/ClientNetworking.cpp]: ... (not noting any more... you get the point... indenting)

File: network/ClientNetworking.h [hunk 8]
```diff
@@ -85,21 +77,21 @@ class ClientNetworking : public std::enable_shared_from_this<ClientNetworking> {
...
     /** Connects to the server on the client's host.  On failure, repeated
         attempts will be made until \a timeout seconds has elapsed. */
-    bool ConnectToLocalHostServer(std::chrono::milliseconds timeout =
+    bool ConnectToLocalHostServer(const std::chrono::milliseconds& timeout =
                                   std::chrono::seconds(10));
 
     /** Sends \a message to the server.  This function actually just enqueues
-        the message for sending and returns immediately. */
-    void SendMessage(Message message);
+        the message for sending a nd returns immediately. */
```
Comment (geoffthemedio) [inline network/ClientNetworking.h]: ?

File: network/ClientNetworking.cpp (hunk 1 above)
Comment (LGM-Doyle) [reply network/ClientNetworking.cpp]: Fixed.

File: network/ClientNetworking.cpp (hunk 2 above)
Comment (LGM-Doyle) [reply network/ClientNetworking.cpp]: No need.  Fixed.

File: network/ClientNetworking.cpp (hunk 4 above)
Comment (LGM-Doyle) [reply network/ClientNetworking.cpp]: fixed.

File: network/ClientNetworking.cpp (hunk 5 above)
Comment (LGM-Doyle) [reply network/ClientNetworking.cpp]: fixed

File: network/ClientNetworking.cpp (hunk 6 above)
Comment (LGM-Doyle) [reply network/ClientNetworking.cpp]: fixed.

File: network/ClientNetworking.cpp (hunk 7 above)
Comment (LGM-Doyle) [reply network/ClientNetworking.cpp]: fixed

File: network/ClientNetworking.h (hunk 8 above)
Comment (LGM-Doyle) [reply network/ClientNetworking.h]: fixed.

**Review by LGM-Doyle — COMMENTED** [review]
All fixed.

@adrianbroher I did not receive an email from AppVeyor.

Also the AppVeyor website does not work on my browser, firefox.  I've enabled all scripting for it, but it just shows three pulsing blue dots.

Travis works correctly.

**Comment by adrianbroher** [comment]
> @adrianbroher I did not receive an email from AppVeyor.

Hm, maybe it's because of the noreply email of GitHub.

> Also the AppVeyor website does not work on my browser, firefox. I've enabled all scripting for it, but it just shows three pulsing blue dots.

Developers lacked permissions until now.  This should be fixed.

File: network/ClientNetworking.h [hunk 9]
```diff
@@ -53,10 +44,11 @@
 class ClientNetworking : public std::enable_shared_from_this<ClientNetworking> {
 public:
     /** The type of list returned by a call to DiscoverLANServers(). */
-    typedef std::vector<std::pair<boost::asio::ip::address, std::string> >  ServerList;
+    using ServerNames =  std::vector<std::string>;
```
Comment (geoffthemedio) [inline network/ClientNetworking.h]: using vs. typedef in this context?

Comment (LGM-Doyle) [reply network/ClientNetworking.h]: In most cases, and this specific case `using` and `typedef` are exactly the same.

tldr; We should replace all of our `typedefs` with `using` declarations.

`using` is more general than `typedef`.  

There is never case where employing `using` produces less functional or more verbose code than `typedef`.  

`using` can be templated and the compiler then knows that the templated thing is a type.  This results in less verbose code and better compiler errors than the typedef alternative.

For example create two templated types, one with `using` and one with `typedef`:
````
template<typename T> using UsingMappedThingType = std::map<T, Thing<T>>;
````
something similar with `typedef` involves nesting in a templated structure.
````
template<typename T> struct StructToHoldTypedef {
    typedef std::map<T, Thing<T>> TypdefMappedThingType;
};
```` 
When further utilizing those types in other templated things, the compiled knows that `UsingMapThingType<T>` is a type, but `StructToHoldTypedef<T>::TypedefMappedThingType` may be a type, member data or a member function, the compile does not know.    

In pre-C++11 code, you can add `typename` decorators to force the compiler to consider this a type, `typename StructToHoldTypedef<T>::TypedefMappedThingType`.  Sometimes adding/removing `typename` causes the compile to fail, because the member you used `typename` for is not actually a type.  In C++11 code, `UsingMappedThingType<T>` in unambiguously a type.

Changing all of the `typedef` statements to `using` declarations would make the code less verbose and more consistent. Keeping `typedef` in some places and `using` in locations where it avoids the use of `typename` adds unnecessary cognitive load.

**Comment by Vezzra** [comment]
Everyone, status report? Anything preventing this from being merged?

**Comment by geoffthemedio** [comment]
````
4>..\..\network\ClientNetworking.cpp(707): error C2039: 'SendMessageW': is not a member of 'ClientNetworking'
4>         c:\users\geoff\desktop\freeorionsdk5\freeorion\network\ClientNetworking.h(44): note: see declaration of 'ClientNetworking'
4>     1>
4>..\..\network\ClientNetworking.cpp(708): error C2065: 'm_impl': undeclared identifier
4>     1>
4>..\..\network\ClientNetworking.cpp(708): error C2227: left of '->SendMessageW' must point to class/struct/union/generic type
4>         ..\..\network\ClientNetworking.cpp(708): note: type is 'unknown-type'
4>     1>
4>..\..\network\ClientNetworking.cpp(710): error C2039: 'GetMessageW': is not a member of 'ClientNetworking'
4>         c:\users\geoff\desktop\freeorionsdk5\freeorion\network\ClientNetworking.h(44): note: see declaration of 'ClientNetworking'
4>     1>
4>..\..\network\ClientNetworking.cpp(711): error C2065: 'm_impl': undeclared identifier
4>     1>
4>..\..\network\ClientNetworking.cpp(711): error C2227: left of '->GetMessageW' must point to class/struct/union/generic type
4>         ..\..\network\ClientNetworking.cpp(711): note: type is 'unknown-type'
````

Same / similar appears in the Appveyor report: https://ci.appveyor.com/project/freeorion/freeorion/build/ci-1088

**Comment by adrianbroher** [comment]
> Anything preventing this from being merged?

Uhm, the Windows CI build is failing as you can see?

**Comment by Vezzra** [comment]
> Uhm, the Windows CI build is failing as you can see?

Oh, sorry... I actually *did* notice that, and then looked at the build log, but not thoroughly - just saw that it ended with:
```
  EmpireWrapper.cpp
  EnumWrapper.cpp
  LoggingWrapper.cpp
  UniverseWrapper.cpp
  DependencyVersions.cpp
Command exited with code 1
```
which looked like the build crashed (maybe because of overtaxing the limited resources of the build environment AppVeyor spawned, I get such crashes on my Mac and Windows systems now and then), as there are no summary lines (detailing how much errors, warnings etc. occured), which I expected though.

Should have taken the two secs to look at the messages a few lines above, then I would have noticed the errors... my bad 😳

**Comment by geoffthemedio** [comment]
To be fair, there are a lot of false positive (false negative?) build failures in the automated test build results.

**Comment by adrianbroher** [comment]
> I actually did notice that, and then looked at the build log, but not thoroughly - just saw that it ended with:

AppVeyor does extract warnings/errors and displays in the `messages` tab right beside the `console` tab so you can get a quick overview what is the exact cause of the issue.  Would be a nice feature for Travis, but as far as I know it doesn't exist.

> which looked like the build crashed (maybe because of overtaxing the limited resources of the build environment AppVeyor spawned, I get such crashes on my Mac and Windows systems now and then)

TravisCI does report build failure (the build broke, marked with a red X) and build errors (the build environment could not be run successfully, e.g compiler crashes, took to long to build, marked with a red !).  I don't know how AppVeyor reports this because I can't remember any breakage so far.

> To be fair, there are a lot of false positive (false negative?) build failures in the automated test build results.

Howso (especially the false positive part)?  I consider the build to be very reliable so far.

**Comment by geoffthemedio** [comment]
> Howso (especially the false positive part)? I consider the build to be very reliable so far.

I (feel like I) often see / get an error about the clang build due to there not being enough time to complete it.

Or here (not sure if it's a legit problem in this specific case):
https://travis-ci.org/freeorion/freeorion/builds/214677710
It can't find `Universe::serialize` apparently?

**Comment by adrianbroher** [comment]
> I (feel like I) often see / get an error about the clang build due to there not being enough time to complete it.

I see your point but I don't agree.  Maybe I can create some statistics about build success rates for the CI in the future.

**Comment by Vezzra** [comment]
Also, that's the Travis CI for OSX. IIRC that hasn't been an issue with AppVeyor, and wrt to the Travis CI build for OSX I *think* @adrianbroher has already done something to get the build times down (and thereby rid of the failed builds due to timeout).

**Comment by adrianbroher** [comment]
What's the whole point of this PR when you add the asio header again to `network/ClientNetworking.h`?  This completely fails the PRs point of isolating the asio dependency inside the PIMPLE implementation.

**Comment by LGM-Doyle** [comment]
@adrianbroher, you are correct for the Windows implementation.  All other OSes work as described above.

Here is the problem with Windows as inferred from the AppVeyor logs and the comment that was already in the code.  (If a real Windows programmer can clarify, please do).

`boost::asio` includes `windows.h`, which defines macros for `min`, `max`, `Message`, `MessageBox` and `PlaySound`.  These conflict with the same named functions in `std` and FreeOrion.  In order to be undefined, they first need to be included.

There is undoubtedly a better way, but I am not compiling on Windows and can only get the compile logs through AppVeyor, which with queue wait times is often hours of round trip time between compiles.  I also have no way to measure the incremental build times, since AppVeyor always builds from scratch. This PR works as described for non-Windows OSes and I am not in a position to efficiently improve the Windows compile times.

**Comment by adrianbroher** [comment]
> @adrianbroher, you are correct for the Windows implementation. All other OSes work as described above.

So it's only 2/3 of the way done or 1/3 still missing, however you prefer to see this.

> If a real Windows programmer can clarify, please do

Yeah here, I'm the author of those lines. What do you want to know? Wait, I just cherry pick the good parts and fix it myself tomorrow.

**Comment by LGM-Doyle** [comment]
@adrianbroher, I am not requesting that anyone do work on my behalf, but thanks for the offer.  I was asking for clarification in case I had inferred a conflicting windows macro when it is a different problem. 

Because I am using AppVeyor as a compiler to get the error log I needed to work back from the broken build to a working build, where I understood what the problem was.

I will try renaming `Message`, `MessageBox` and `PlaySound` so that there are no future conflicts with the Windows `windows.h` macros.

We can't rename `std::min` and `std::max`, so I will find another workaround.

I've marked the PR WIP, and will leave it that way until Ii think that the Windows portion is done.

**Comment by Vezzra** [comment]
If fixing these problems takes longer, can we reassign to the "optional for 0.4.7" milestone? This PR isn't really *critical* for the relase, after all...

**Comment by LGM-Doyle** [comment]
I meant to do that when I changed it to WIP.

**Comment by o01eg** [comment]
What if isolate the code uses "windows.h"?

**Comment by LGM-Doyle** [comment]
@o01eg that is what I have done in the current PR.  The objection is that it makes the incremental build performance of the Windows version of `ClientNetworking.h` worse than the other OSes.

The problem is that the `windows.h` is being implicitly included in headers from `boost` libraries that we do not control.

**Comment by adrianbroher** [comment]
Merged as f731059e293e1d960490dcac1759a515be50318c, leaving out most recent commits.

**Comment by LGM-Doyle** [comment]
Thanks.
